modules/indexing/src/test/java/org/apache/ignite/util/GridCommandHandlerIndexForceRebuildTest.java

```diff
@@ -0,0 +1,428 @@
+            "  IndexRebuildStatusInfoContainer [groupName=" + GRP_NAME_2 + ", cacheName=" + CACHE_NAME_2_1 + "]"));
+
+        assertEquals("Unexpected number of lines in output.", 20, outputStr.split("\n").length);
+    }
+
+    /**
+     * Triggers indexes rebuild for ALL caches on grid node with index {@code igniteIdx}.
+     *
+     * @param igniteIdx Node index.
+     * @param excludedCacheNames Collection of cache names for which
+     *  end of index rebuilding is not awaited.
+     * @throws Exception if failed.
+     */
+    private void triggerIndexRebuild(int igniteIdx, Collection<String> excludedCacheNames) throws Exception {
+        stopGrid(igniteIdx);
+
+        GridTestUtils.deleteIndexBin(getTestIgniteInstanceName(2));
+
+        GridQueryProcessor.idxCls = BlockingIndexing.class;
+        final IgniteEx ignite = startGrid(igniteIdx);
+
+        resetBaselineTopology();
+        awaitPartitionMapExchange();
+        waitForIndexesRebuild(ignite, 30_000, excludedCacheNames);
+    }
+
+    /** */
+    private boolean waitForIndexesRebuild(IgniteEx ignite) throws IgniteInterruptedCheckedException {
+        return waitForIndexesRebuild(ignite, 30_000, Collections.emptySet());
+    }
+
+    /**
+     * @param ignite Ignite instance.
+     * @param timeout timeout
+     * @param excludedCacheNames Collection of cache names for which
+     *  end of index rebuilding is not awaited.
+     * @return {@code True} if index rebuild was completed before {@code timeout} was reached.
+     * @throws IgniteInterruptedCheckedException if failed.
+     */
+    private boolean waitForIndexesRebuild(IgniteEx ignite, long timeout, Collection<String> excludedCacheNames)
+        throws IgniteInterruptedCheckedException
+    {
+        return GridTestUtils.waitForCondition(
+            () -> ignite.context().cache().publicCaches()
+                .stream()
+                .filter(c -> !excludedCacheNames.contains(c.getName()))
+                .allMatch(c -> c.indexReadyFuture().isDone()),
+            timeout);
+    }
+
+    /**
+     * @param ignite IgniteEx instance.
+     * @param cacheName Name of checked cache.
+     * @return newly installed LogListener.
+     */
+    private LogListener installRebuildCheckListener(IgniteEx ignite, String cacheName) {
+        final MessageOrderLogListener lsnr = new MessageOrderLogListener(
+            new MessageOrderLogListener.MessageGroup(true)
+                .add("Started indexes rebuilding for cache \\[name=" + cacheName + ".*")
+                .add("Finished indexes rebuilding for cache \\[name=" + cacheName + ".*")
+        );
+
+        ListeningTestLogger impl = GridTestUtils.getFieldValue(ignite.log(), "impl");
+        assertNotNull(impl);
+
+        impl.registerListener(lsnr);
+
+        return lsnr;
+    }
+
+    /** */
+    private void removeLogListener(IgniteEx ignite, LogListener lsnr) {
```

Comment:
Lets store `ListeniningTestLogger` in an instance field instead of reflection.

Reply:
I tried. Unfortunately it doesn't work

Reply:
I think you must find out the problem. You can ask me or other teammates for any assistance.